rhino/src/main/java/org/mozilla/javascript/SlotMapOwner.java

```diff
@@ -0,0 +1,60 @@
+package org.mozilla.javascript;
+
+import java.lang.invoke.MethodHandles;
+import java.lang.invoke.VarHandle;
+
+public abstract class SlotMapOwner {
+    private static final long serialVersionUID = 1L;
+
+    /**
+     * This holds all the slots. It may or may not be thread-safe, and may expand itself to a
+     * different data structure depending on the size of the object.
+     */
+    private volatile SlotMap slotMap;
+
+    protected SlotMapOwner() {
+        slotMap = createSlotMap(0);
+    }
+
+    protected SlotMapOwner(int capacity) {
+        slotMap = createSlotMap(capacity);
+    }
+
+    protected SlotMapOwner(SlotMap map) {
+        slotMap = map;
+    }
+
+    protected static SlotMap createSlotMap(int initialSize) {
+        Context cx = Context.getCurrentContext();
+        if ((cx != null) && cx.hasFeature(Context.FEATURE_THREAD_SAFE_OBJECTS)) {
+            return new ThreadSafeSlotMapContainer(initialSize);
+        } else if (initialSize == 0) {
+            return SlotMapContainer.EMPTY_SLOT_MAP;
+        } else if (initialSize > SlotMapContainer.LARGE_HASH_SIZE) {
+            return new HashSlotMap();
+        } else {
+            return new EmbeddedSlotMap();
+        }
+    }
+
+    @SuppressWarnings("AndroidJdkLibsChecker")
+    private static final VarHandle SLOT_MAP = getSlotMapHandle();
+
+    @SuppressWarnings("AndroidJdkLibsChecker")
+    private static VarHandle getSlotMapHandle() {
+        try {
+            return MethodHandles.lookup()
```

Comment:
Is there a way for us to verify whether this feature is going to break Android compatibility? I think that our hypothesis is that it won't affect Android users unless they try to use "thread safe slot maps." However I don't know whether the incompatibility would be triggered by using the VarHandle, or merely by trying to load the class. Is there any way that we can verify this?

Reply:
On Android 33 (Tiramisu) or later (released in 2022) `VarHandle`s are supported so this should work fine. I am not sure what the behaviour is like before that (whether the class is present but returns an error, or if the app will fail at install or build time). Since we already suppressed warnings on `ThreadSafeSlotMapContainer` for android compatibility concerns would it be reasonable to move the VarHandle and uses to a separate class only used by the thread safe maps?

Reply:
I would be a lot more comfortable if the VarHandle stuff could be isolated to a class that's not loaded in the non-thread-safe case. It's too bad we don't have a clear compatibility policy for Android -- I've noticed that people have lots of versions that they use and it's hard to predict what will break.

Reply:
Separated the multithreaded case to a separate class, and moved that to the later PR for thread safe stuff. I've kept it as a class in the same file so that it can have private access to the field.